pytential/qbx/geometry.py

```diff
@@ -797,52 +772,49 @@ def user_target_to_center(self):
+        result = actx.empty(
+                target_info.ntargets, tgt_assoc_result.target_to_center.dtype
+                )
         result[:self.ncenters].fill(target_state.NO_QBX_NEEDED)
         result[self.ncenters:] = tgt_assoc_result.target_to_center
 
-        return result.with_queue(None)
+        return actx.freeze(result)
 
     @memoize_method
     @log_process(logger)
     def center_to_tree_targets(self):
         """Return a :class:`CenterToTargetList`. See :meth:`user_target_to_center`
         for the reverse look-up table with targets in user order.
-
         |cached|
         """
 
+        actx = self._setup_actx
         user_ttc = self.user_target_to_center()
 
-        with cl.CommandQueue(self.cl_context) as queue:
-            tree_ttc = cl.array.empty_like(user_ttc).with_queue(queue)
-            tree_ttc[self.tree().sorted_target_ids] = user_ttc
-
-            filtered_tree_ttc = cl.array.empty(queue, tree_ttc.shape, tree_ttc.dtype)
-            filtered_target_ids = cl.array.empty(
-                    queue, tree_ttc.shape, tree_ttc.dtype)
-            count = cl.array.empty(queue, 1, tree_ttc.dtype)
+        tree_ttc = actx.empty_like(user_ttc)
+        tree_ttc[self.tree().sorted_target_ids] = user_ttc
 
-            self.code_getter.filter_center_and_target_ids(tree_ttc.dtype)(
-                    tree_ttc, filtered_tree_ttc, filtered_target_ids, count,
-                    queue=queue, size=len(tree_ttc))
+        filtered_tree_ttc = actx.empty(tree_ttc.shape, dtype=tree_ttc.dtype)
+        filtered_target_ids = actx.empty(tree_ttc.shape, dtype=tree_ttc.dtype)
+        count = actx.empty(1, dtype=tree_ttc.dtype)
 
-            count = count.get().item()
+        self.code_getter.filter_center_and_target_ids(tree_ttc.dtype)(
+                tree_ttc, filtered_tree_ttc, filtered_target_ids, count,
+                queue=actx.queue, size=len(tree_ttc))
 
-            filtered_tree_ttc = filtered_tree_ttc[:count]
-            filtered_target_ids = filtered_target_ids[:count].copy()
+        count = actx.to_numpy(count).item()
+        filtered_tree_ttc = filtered_tree_ttc[:count]
+        filtered_target_ids = actx.np.copy(filtered_target_ids[:count])
 
-            center_target_starts, targets_sorted_by_center, _ = \
-                    self.code_getter.key_value_sort(queue,
-                            filtered_tree_ttc, filtered_target_ids,
-                            self.ncenters, tree_ttc.dtype)
+        center_target_starts, targets_sorted_by_center, _ = \
+                self.code_getter.key_value_sort(actx.queue,
+                        filtered_tree_ttc, filtered_target_ids,
+                        self.ncenters, tree_ttc.dtype)
 
-            result = CenterToTargetList(
-                    starts=center_target_starts,
-                    lists=targets_sorted_by_center).with_queue(None)
-
-            return result
+        actx.queue.finish()
```

Comment:
This seems to be introducing some synchronization issues. Most of the functions in `QBXFMMGeometryData` were creating their own queues to precompute data and this PR switched them to using the array context from the constructor and freezing the results.

Not sure why that would cause issues (it doesn't seem like anything is returned thawed), but it does :confounded:

Reply:
After some playing around with this: it was only causing failures when using `fmmlib` where it was using a
```
with cl.CommandQueue() as queue:
	geo_data = ToHostTransferredGeoDataWrapper(queue, geo_data)
```
Switching that to using the array context inside `geo_data` seems to fix any failures. It still seems like something is a bit brittle though :(